feat: add package model

nyx-lyb3ra opened the pull request:
This is completely WIP atm. Just pushing to get some feedback

nyx-lyb3ra committed:
work in progress

dar5hak commented on src/models/Package.vala:
```diff
@@ -0,0 +1,165 @@
+    }
+}
+
+class Vamp.Repository : Object {
```
Should this (and others) not extend `Json.Serializable`, given that we are overriding the methods?

dar5hak replied:
Also, shall we keep all of these in the top-level namespace? Would it be an overkill to separate them into a `Vamp.Package` namespace?

colinkiama replied:
I think that separating them into another namespace would be a good idea however, I can't come up with any advantages for doing it right now.

Maybe it's best to keep a note of this suggestion for now?

nyx-lyb3ra replied:
> Should this (and others) not extend Json.Serializable, given that we are overriding the methods?

Repository is not overwriting any method. `to_json()` does not belong to the interface, and `from_json()` is a static method: static methods cannot be inherited in any way, and it does not belong to `JsonSerializable` either.

Both are methods that I simply implemented for convenience when working with them.

nyx-lyb3ra replied:
> Also, shall we keep all of these in the top-level namespace? Would it be an overkill to separate them into a Vamp.Package namespace?

Well, at this point I wouldn't know much what to answer. My way of working with Vala is to just have 1 namespace and that's it, mostly because my code usually tries to be compatible with C and GObject Introspection and:
1. In C, it is not comfortable to have long namespaces, especially because you will have to write them all the time (see `gtk_window_present()`, `soup_session_send_and_read_async()`, etc).
2. GObject Introspection does not support nested namespaces, so if you create them in Vala, for Python, JS, Rust, and other languages they will become part of the class name (for example, `Vamp.Package.Repository` in Python becomes `Vamp.PackageRepository`).
3. Many Vala projects usually ask you not to use the `using` keyword, but to be explicit, so we are back to the annoyances of point 1.

In fact, I was considering dropping namespaces entirely and only using them if we are going to expose a library, and even then, using them only in classes/methods that will be exposed as public API.

nyx-lyb3ra replied:
Although my position on this is not definitive, I just stated the way I work. If you really believe that we should take advantage of the use of namespaces (although I really don't know what real advantage they have in Vala), then I will use them.

dar5hak replied:
Fair enough. I suggested it just for cognitive ease, but if it makes accessing more complex, let's just stick to this. I don't see us exposing a library any time soon, and if we do, that's a problem for another day.

dar5hak commented:
The general approach LGTM.

devnix commented on src/models/Package.vala:
```diff
@@ -0,0 +1,165 @@
+    public static Person from_json (Json.Node node) {
+        assert (node.get_node_type () == OBJECT);
+        return (Person) Json.gobject_deserialize (typeof (Person), node);
+    }
+
+    public Json.Node to_json () {
+        return Json.gobject_serialize (this);
+    }
+}
+
+class Vamp.Bugs : Object {
+    public string url { get; set; }
+    public string email { get; set; }
+
+    public static Bugs from_json (Json.Node node) {
+        assert (node.get_node_type () == OBJECT);
+        return (Bugs) Json.gobject_deserialize (typeof (Bugs), node);
+    }
+
+    public Json.Node to_json () {
+        return Json.gobject_serialize (this);
+    }
+}
+
+Gee.List<string> string_list_from_json (Json.Node node) {
+    assert (node.get_node_type () == ARRAY);
+
+    var array = node.get_array ();
+    var result = new Gee.ArrayList<string> ();
+
+    array.foreach_element ((_, __, element_node) => {
+        if (element_node.get_value_type () == Type.STRING) {
+            return;
+        }
+
+        result.add(element_node.get_string ());
+    });
+
+    return result;
+}
```
I think we should have one file per public class: https://docs.elementary.io/develop/writing-apps/code-style#classes-and-files

Also: we will have this classes on the `Vamp` namespace or in a more specific namespace?

nyx-lyb3ra replied:
As I said in another comment, since Vala namespaces have no real advantages (literally not even in terms of member accessibility), my idea was simply not to use them unless strictly necessary: for example, if we are exposing a public API or if we want to simulate static classes.

devnix reviewed (commented):
In general, it looks great!

devnix commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "editor.rulers": [80]
+}
```
What do you think about this? Should we leave editor-specific files outside of the repository, or is it OK to commit them? I'm personally against it in favor of having a `.editorconfig` file, besides the fact it can't add rulers.

dar5hak replied:
From past experience, I'm in favour of adding settings which benefit everyone (esp. when they aren't covered in `.editorconfig`), but a ruler probably makes more sense in @nahuelwexd's global settings than in this file here. It's merely visual.

nyx-lyb3ra replied:
This was because I ran `git add .` on the project without taking a good look at what I was committing. I will fix it later

dar5hak committed:
fix: lint errors

dar5hak commented:
Great! vala-lint thinks those parentheses inside the regex are method calls!

devnix commented:
> Great! vala-lint thinks those parentheses inside the regex are method calls!

Oomph! Should we open a bug report?

dar5hak commented:
There's already an open issue: https://github.com/vala-lang/vala-lint/issues/141

dar5hak committed:
Disable space-before-paren on the regex

Until https://github.com/vala-lang/vala-lint/issues/141 is fixed

nyx-lyb3ra committed:
remove .vscode directory

nyx-lyb3ra committed:
move Person parsing logic to its own method

nyx-lyb3ra committed:
add a missing check

nyx-lyb3ra committed:
fix the missing check

nyx-lyb3ra committed:
some more progress

colinkiama commented:
> There's already an open issue: https://github.com/vala-lang/vala-lint/issues/141

They do take ages to merge in changes (doesn't seem to have dedicated maintainers) so lets just ignore the problem for now.

The rule that's being broken is the `space-before-paren`.

We can make vala-lint ignore the line containing the regex by adding this at the end of the line: `// vala-lint=space-before-paren, line-length`.

dar5hak commented:
@colinkiama Yeah, we ignored `space-before-paren` already in https://github.com/dar5hak/vamp/pull/9/commits/3a7f9d13664763cf182ad5ef10d538e5777af7ef. Just not `line-length`.

colinkiama committed:
Start working on file test

colinkiama committed:
Test package config file can be parsed

colinkiama committed:
Add references to npm and drakkar package config specs

colinkiama committed:
Full package config file test

colinkiama committed:
Fix: HashMap properties failing to be deserialized

colinkiama committed:
Fix: string_list_from_json method not adding string values

colinkiama committed:
Add "equals" methods to package models

colinkiama committed:
Check if contributors are parsed correctly

colinkiama committed:
Handle parsing of "type" property names

colinkiama committed:
Fix: Respository property object parsing

The parse result for `repository` was set to "false" even when parsing was successful

colinkiama committed:
Test full package configuration file deserialization

colinkiama committed:
Serialized keywords

colinkiama committed:
Fix serialization of "type" properties

colinkiama committed:
Serialize "dependencies" properties

colinkiama committed:
Compare serialized content with expecrted content

colinkiama committed:
Move File test to unit tests file

colinkiama committed:
refactor: Move public model classes into their own files

colinkiama committed:
test config vapi formatting

colinkiama committed:
Formatting

colinkiama committed:
Fix: Missing config header for tests

colinkiama committed:
Replaced assertcmp methods with assert

colinkiama committed:
Explicit disable new lines from being added to test asset files

colinkiama committed:
Rename `equals` methods with `equal`

colinkiama committed:
Add space between cases

colinkiama committed:
Packages and generators in unit tests now initialised with object initialisers

colinkiama committed:
Fix "desrialise_package_config" typo

colinkiama committed:
Stop checking for errors from boolean value of `FileUtils.get_contents`

colinkiama committed:
Reduce usage of CCode Attributes in config.vapi

colinkiama committed:
Remove double blank line

colinkiama committed:
Use literal map in for test configuration files

colinkiama committed:
Split project into `cli` and `vamp` directories

colinkiama committed:
Merge literal strings in test build file to one string

colinkiama committed:
Using full qualifier names for JSON NodeType enums

nyx-lyb3ra commented:
Going to do some changes before merging
